File: src/main/java/com/github/svegon/capi/event/network/S2CLoginPacketListener.java
package com.github.svegon.capi.event.network;

import com.github.svegon.capi.event.ListenerCollectionFactory;
import com.github.svegon.capi.event.ListenerList;
import com.github.svegon.capi.event.ListenerSet;
import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;
import net.minecraft.network.ClientConnection;
import net.minecraft.network.listener.ClientLoginPacketListener;
import net.minecraft.network.packet.Packet;
import net.minecraft.network.packet.s2c.login.*;
import net.minecraft.text.Text;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;

public interface S2CLoginPacketListener extends ClientLoginPacketListener {
    Event<S2CLoginPacketListener> CLIENT_PACKET_SENT_EVENT = EventFactory.createArrayBacked(
            S2CLoginPacketListener.class,
            (listeners) -> new S2CLoginPacketListener() {
                @Override
                public void apply(Packet<ClientLoginPacketListener> packet, CallbackInfo info) {
                    for (S2CLoginPacketListener listener : listeners) {
                        listener.apply(packet, info);

                        if (info.isCancelled()) {
                            return;
                        }
                    }
                }
            });
    ListenerList<S2CLoginPacketListener> LISTENER_LIST = ListenerCollectionFactory.listenersVector(
            (l) -> new S2CLoginPacketListener() {
                @Override
                public void apply(Packet<ClientLoginPacketListener> packet, CallbackInfo info) {
                    for (S2CLoginPacketListener listener : l) {
                        listener.apply(packet, info);

                        if (info.isCancelled()) {
                            return;
                        }
                    }
                }
            });
    ListenerSet<S2CLoginPacketListener> LISTENER_SET = ListenerCollectionFactory.listenersLinkedSet(
            (s) -> new S2CLoginPacketListener() {
                @Override
                public void apply(Packet<ClientLoginPacketListener> packet, CallbackInfo info) {
                    for (S2CLoginPacketListener listener : s) {
                        listener.apply(packet, info);

                        if (info.isCancelled()) {
                            return;
                        }
                    }
                }
            });

    default void apply(Packet<ClientLoginPacketListener> packet, CallbackInfo callback) {
        packet.apply(this);
    }

    @Override
    default boolean isConnectionOpen() {
        return false;
    }

    @Override
    default void onHello(LoginHelloS2CPacket packet) {

    }

    @Override
    default void onSuccess(LoginSuccessS2CPacket packet) {

    }

    @Override
    default void onDisconnect(LoginDisconnectS2CPacket packet) {

    }

    @Override
    default void onCompression(LoginCompressionS2CPacket packet) {

    }

    @Override
    default void onQueryRequest(LoginQueryRequestS2CPacket packet) {

    }

    @Override
    default void onDisconnected(Text reason) {

    }
}

File: src/main/kotlin/io/github/svegon/mclientapi/event/network/S2CLoginPacketListener.kt
package io.github.svegon.mclientapi.event.network

import net.minecraft.network.PacketListener
import net.minecraft.network.protocol.Packet
import net.minecraft.network.protocol.cookie.ClientboundCookieRequestPacket
import net.minecraft.network.protocol.login.ClientLoginPacketListener
import net.minecraft.network.protocol.login.ClientboundCustomQueryPacket
import net.minecraft.network.protocol.login.ClientboundHelloPacket
import net.minecraft.network.protocol.login.ClientboundLoginCompressionPacket
import net.minecraft.network.protocol.login.ClientboundLoginDisconnectPacket
import net.minecraft.network.protocol.login.ClientboundLoginFinishedPacket
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo
import java.util.function.Function

interface S2CLoginPacketListener : S2CPacketListener, ClientLoginPacketListener {
    override fun handleHello(packet: ClientboundHelloPacket) {}

    override fun handleLoginFinished(packet: ClientboundLoginFinishedPacket) {}

    override fun handleDisconnect(packet: ClientboundLoginDisconnectPacket) {}

    override fun handleCompression(packet: ClientboundLoginCompressionPacket) {}

    override fun handleCustomQuery(packet: ClientboundCustomQueryPacket) {}

    override fun handleRequestCookie(packet: ClientboundCookieRequestPacket) {}

    object EmptyInvoker : S2CLoginPacketListener

    object InvokerFactory : Function<Array<S2CLoginPacketListener>, S2CLoginPacketListener> {
        override fun apply(listeners: Array<S2CLoginPacketListener>): S2CLoginPacketListener {
            return object : S2CLoginPacketListener {
                override fun intercept(packet: Packet<out PacketListener>, callback: CallbackInfo) {
                    for (listener in listeners) {
                        listener.intercept(packet, callback)

                        if (callback.isCancelled) {
                            return
                        }
                    }
                }
            }
        }
    }
}